### PR TITLE
Allow all supported platforms & extensions

### PR DESCRIPTION
**Pull request by jpeg-ms**
This adds explicit support for all known platforms and file extensions currently supported by App Center.
Alternatively, we could add an option to bypass checks and let the API itself accept or reject requests.

This is meant to address issue #136

### CODE REVIEW

**Comment by jerietve** [comment]
In my opinion the copying of this logic is getting out of hand. It's in the portal, in the releases service in multiple places, in the CLI, in FastLane. Perhaps we're complicating things too much just to provide errors without roundtrips.

File: README.md [hunk 1]
```diff
@@ -83,15 +83,12 @@ Here is the list of all existing parameters:
...
+| `owner_name` <br/> `APPCENTER_OWNER_NAME` | Owner name as found in the App's URL in App Center |
 | `app_name` <br/> `APPCENTER_APP_NAME` | App name as found in the App's URL in App Center, if there is no app with such name, you will be prompted to create one |
 | `app_display_name` <br/> `APPCENTER_APP_DISPLAY_NAME` | App display name to use when creating a new app |
 | `app_os` <br/> `APPCENTER_APP_OS` | App OS. Used for new app creation, if app 'app_name' was not found |
 | `app_platform` <br/> `APPCENTER_APP_PLATFORM` | App Platform. Used for new app creation, if app 'app_name' was not found |
-| `apk` <br/> `APPCENTER_DISTRIBUTE_APK` | Build release path for android build |
-| `aab` <br/> `APPCENTER_DISTRIBUTE_AAB` | Build release path for android app bundle build |
-| `ipa` <br/> `APPCENTER_DISTRIBUTE_IPA` | Build release path for iOS builds |
-| `file` <br/> `APPCENTER_DISTRIBUTE_FILE` | Build release path for generic builds (.aab, .app, .app.zip, .apk, .dmg, .ipa, .pkg) |
+| `file` <br/> `APPCENTER_DISTRIBUTE_FILE` | Build release path for generic builds |
```
Comment (jerietve) [inline README.md]: 'for generic builds' can now go

File: lib/fastlane/plugin/appcenter/actions/appcenter_upload_action.rb [hunk 2]
```diff
@@ -2,6 +2,18 @@ module Fastlane
   module Actions
     module Constants
       MAX_RELEASE_NOTES_LENGTH = 5000
+      SUPPORTED_EXTENSIONS = {
+          android: %w(.aab .apk),
+          ios: %w(.ipa),
+          mac: %w(.app .app.zip .dmg .pkg),
```
Comment (jerietve) [inline lib/fastlane/plugin/appcenter/actions/appcenter_upload_action.rb]: The portal does not have .app on the support list. Whether it is required to be ".app.zip" or any .zip is fine also depends on the implementation of this code; it's handled differently in different places.

File: lib/fastlane/plugin/appcenter/actions/appcenter_upload_action.rb [hunk 3]
```diff
@@ -131,15 +143,18 @@ def self.run_release_upload(params)
...
-          UI.user_error!("Can't distribute #{file_ext} to stores, please use `destination_type: 'group'`") if %w(.app .app.zip .dmg .pkg).include? file_ext
+          UI.user_error!("Can't distribute #{file_ext} to stores, please use `destination_type: 'group'`") unless Constants::STORE_SUPPORTED_EXTENSIONS.include? file_ext
         end
 
         unless params[:file].to_s.empty?
-          if %w[.dmg .pkg].include? file_ext
+          if Constants::FULL_VERSION_REQUIRED_EXTENSIONS.include? file_ext
             UI.user_error!("Fields `version` and `build_number` must be specified to upload a #{file_ext} file") if build_number.to_s.empty? || version.to_s.empty?
             release_upload_body = { build_version: version, build_number: build_number }
+          elsif Constants::VERSION_REQUIRED_EXTENSIONS.include? file_ext
```
Comment (jerietve) [inline lib/fastlane/plugin/appcenter/actions/appcenter_upload_action.rb]: file_ext will probably contain ".zip" for ".app.zip" files. This requirement of having a version number should be OS dependent as well.

File: lib/fastlane/plugin/appcenter/actions/appcenter_upload_action.rb [hunk 4]
```diff
@@ -297,7 +313,7 @@ def self.available_options
 
           FastlaneCore::ConfigItem.new(key: :app_name,
                                   env_name: "APPCENTER_APP_NAME",
-                               description: "App name as found in the App's URL in App Center. If there is no app with such name, you will be prompted to create one",
+                               description: "App name as found in the App's URL in App Center, if there is no app with such name, you will be prompted to create one",
```
Comment (jerietve) [inline lib/fastlane/plugin/appcenter/actions/appcenter_upload_action.rb]: The period looks more grammatically correct.

File: README.md (hunk 1 above)
Comment (jpeg-ms) [reply README.md]: Oops I forgot to re-update the doc after changing in the code

File: lib/fastlane/plugin/appcenter/actions/appcenter_upload_action.rb (hunk 2 above)
Comment (jpeg-ms) [reply lib/fastlane/plugin/appcenter/actions/appcenter_upload_action.rb]: The portal supports .app and automatically zips for you.

File: lib/fastlane/plugin/appcenter/actions/appcenter_upload_action.rb (hunk 3 above)
Comment (jpeg-ms) [reply lib/fastlane/plugin/appcenter/actions/appcenter_upload_action.rb]: Good catch!

File: lib/fastlane/plugin/appcenter/actions/appcenter_upload_action.rb [hunk 5]
```diff
@@ -554,7 +578,7 @@ def self.example_code
             api_token: "...",
             owner_name: "appcenter_owner",
             app_name: "testing_app",
-            aab: "./app.aab",
+            file: "./app.aab",
             destinations: "Alpha",
             destination_type: "store",
             build_number: "3",
```
Comment (jpeg-ms) [inline lib/fastlane/plugin/appcenter/actions/appcenter_upload_action.rb]: Looking at this now I can see that the examples don't make much sense. aab doesn't need build_number, nor does apk.

File: lib/fastlane/plugin/appcenter/actions/appcenter_upload_action.rb (hunk 4 above)
Comment (jpeg-ms) [reply lib/fastlane/plugin/appcenter/actions/appcenter_upload_action.rb]: Right, this was meant to unify README and this (they should be copy-paste)

File: lib/fastlane/plugin/appcenter/actions/appcenter_upload_action.rb (hunk 2 above)
Comment (jpeg-ms) [reply lib/fastlane/plugin/appcenter/actions/appcenter_upload_action.rb]: Nevermind my comment. macOS does this for you (at least Chrome on macOS).
But anyway the plugin does this automatic zipping for you :)

**Comment by Ajaykn21** [comment]
> In my opinion the copying of this logic is getting out of hand. It's in the portal, in the releases service in multiple places, in the CLI, in FastLane. 

I feel the same, and for instance, we don't allow .aab for all the stores as well (for Android-Intune, this should throw an error).
And In future, we should be able to handle only at Portal and API level for all these validations, so that other clients like cli, fast lane, etc, just depends on API.

Whats your thoughts on this ? @jpeg-ms

**Comment by jpeg-ms** [comment]
> > In my opinion the copying of this logic is getting out of hand. It's in the portal, in the releases service in multiple places, in the CLI, in FastLane.
> 
> I feel the same, and for instance, we don't allow .aab for all the stores as well (for Android-Intune, this should throw an error).
> And In future, we should be able to handle only at Portal and API level for all these validations, so that other clients like cli, fastlane, etc, just depends on API.
> 
> Whats your thoughts on this ? @jpeg-ms

Yeah I also tend to agree with this. See my initial comment:
> Alternatively, we could add an option to bypass checks and let the API itself accept or reject requests.
We could modify this PR to show warnings only, or refactor the entire logic to ignore these filters and try to do API calls without any initial check.

File: README.md [hunk 6]
```diff
@@ -83,15 +83,12 @@ Here is the list of all existing parameters:
...
+| `owner_name` <br/> `APPCENTER_OWNER_NAME` | Owner name as found in the App's URL in App Center |
+| `app_name` <br/> `APPCENTER_APP_NAME` | App name as found in the App's URL in App Center. If there is no app with such name, you will be prompted to create one |
 | `app_display_name` <br/> `APPCENTER_APP_DISPLAY_NAME` | App display name to use when creating a new app |
 | `app_os` <br/> `APPCENTER_APP_OS` | App OS. Used for new app creation, if app 'app_name' was not found |
 | `app_platform` <br/> `APPCENTER_APP_PLATFORM` | App Platform. Used for new app creation, if app 'app_name' was not found |
-| `apk` <br/> `APPCENTER_DISTRIBUTE_APK` | Build release path for android build |
-| `aab` <br/> `APPCENTER_DISTRIBUTE_AAB` | Build release path for android app bundle build |
-| `ipa` <br/> `APPCENTER_DISTRIBUTE_IPA` | Build release path for iOS builds |
-| `file` <br/> `APPCENTER_DISTRIBUTE_FILE` | Build release path for generic builds (.aab, .app, .app.zip, .apk, .dmg, .ipa, .pkg) |
+| `file` <br/> `APPCENTER_DISTRIBUTE_FILE` | File path to the release build to publish |
```
Comment (Ajaykn21) [inline README.md]: This is cool..

File: lib/fastlane/plugin/appcenter/actions/appcenter_upload_action.rb [hunk 7]
```diff
@@ -336,52 +357,59 @@ def self.available_options
                               verify_block: proc do |value|
                                 accepted_formats = [".apk"]
                                 file_extname_full = Helper::AppcenterHelper.file_extname_full(value)
-                                UI.user_error!("Only \".apk\" formats are allowed, you provided \"#{file_extname_full}\"") unless accepted_formats.include? file_extname_full
+                                self.optional_error("Only \".apk\" formats are allowed, you provided \"#{file_extname_full}\"") unless accepted_formats.include? file_extname_full
                               end),
 
           FastlaneCore::ConfigItem.new(key: :aab,
```
Comment (Ajaykn21) [inline lib/fastlane/plugin/appcenter/actions/appcenter_upload_action.rb]: I assume we are keeping this now for backward compatibility ?
Whats the plan to remove this ?

Comment (jpeg-ms) [reply lib/fastlane/plugin/appcenter/actions/appcenter_upload_action.rb]: No plan to remove. Just marked as deprecated. :)

Comment (jpeg-ms) [reply lib/fastlane/plugin/appcenter/actions/appcenter_upload_action.rb]: To be clear, to remove we would have to change the major version number

**Pull request merged**